common/web/gesture-recognizer/src/inputEventEngine.ts

```diff
@@ -18,20 +24,77 @@ namespace com.keyman.osk {
     abstract registerEventHandlers();
     abstract unregisterEventHandlers();
 
-    onInputStart(coord: InputEventCoordinate) {
-      this.emit(InputEventEngine.INPUT_UPDATE_EVENT_NAME, TrackedInputState.START, coord);
+    /**
+     * @param identifier The identifier number corresponding to the input sequence.
+     */
+    hasActiveSequence(identifier: number) {
+      return this._activeSequenceWrappers.findIndex((seq) => seq.item.identifier == identifier) != -1;
+    }
+
+    private getSequenceWrapperWithId(identifier: number) {
+      return this._activeSequenceWrappers.find((seq) => seq.item.identifier == identifier);
+    }
+
+    public cleanupSequenceWithId(identifier: number) {
+      this._activeSequenceWrappers = this._activeSequenceWrappers.filter((seq) => seq.item.identifier != identifier);
     }
 
-    onInputMove(coord: InputEventCoordinate) {
-      this.emit(InputEventEngine.INPUT_UPDATE_EVENT_NAME, TrackedInputState.MOVE, coord);
+    protected buildSampleFor(clientX, clientY): InputSample {
+      const targetRect = this.config.targetRoot.getBoundingClientRect();
+      return {
+        clientX: clientX,
+        clientY: clientY,
+        targetX: clientX - targetRect.left,
+        targetY: clientY - targetRect.top,
+        t: performance.now()
+      };
     }
 
-    onInputMoveCancel(coord: InputEventCoordinate) {
-      this.emit(InputEventEngine.INPUT_UPDATE_EVENT_NAME, TrackedInputState.CANCEL, coord);
+    onInputStart(identifier: number, sample: InputSample, target: EventTarget) {
+      let sequence = new InputSequence(identifier, target, this instanceof TouchEventEngine);
+      sequence.addSample(sample);
+
+      let sequenceWrapper = new Incomplete<InputSequence, InputSample>(sequence);
+
+      // External objects may desire to directly terminate handling of
+      // input sequences under specific conditions.
+      let _this = this;
+      sequenceWrapper.on('cancel', function() {
+        _this.cleanupSequenceWithId(identifier);
+      });
+
+      sequenceWrapper.on('end', function() {
+        _this.cleanupSequenceWithId(identifier);
+      });
+
+      this._activeSequenceWrappers.push(sequenceWrapper);
+      this.emit(InputEventEngine.INPUT_START_EVENT_NAME, sequenceWrapper);
     }
 
-    onInputEnd(coord: InputEventCoordinate) {
-      this.emit(InputEventEngine.INPUT_UPDATE_EVENT_NAME, TrackedInputState.END, coord);
+    onInputMove(identifier: number, sample: InputSample) {
+      const sequenceWrapper = this.getSequenceWrapperWithId(identifier);
+      sequenceWrapper.item.addSample(sample);
```

Comment:
```suggestion
      if(!sequenceWrapper) {
        // TODO: log unexpected?
        return;
      }
      sequenceWrapper.item.addSample(sample);
```

Given you are always verifying the existence of `sequenceWrapper` elsewhere, you should probably be doing the same here!